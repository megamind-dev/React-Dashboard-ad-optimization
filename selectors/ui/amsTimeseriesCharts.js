import createCachedSelector from 're-reselect'
import isEmpty from 'lodash/isEmpty'
import zipObject from 'lodash/zipObject'
import get from 'lodash/get'

import moment from 'utilities/moment'
import { getMetric, aggregateMetricByDateGroups } from 'helpers/metrics'
import { formatNumber, formatDate } from 'helpers/formatting'
import { FACT_TYPES } from 'constants/filters'
import {
    LEFT_AXIS_FIRST,
    RIGHT_AXIS,
    CHART_METRIC_KEYS,
} from 'constants/charts'
import { FACT_TYPE_KEYS } from 'constants/factTypes'
import { FILTERS, CHARTS } from 'constants/reducerKeys'

import { selectDomainState, selectDomainValue } from './ui'
import { selectPageAggregation } from './filters'

export const selectChartMetrics = createCachedSelector(
    selectDomainState,
    (state, pageName) => pageName,
    (state, pageName, chartName) => chartName,
    (domainState, pageName, chartName) =>
        get(domainState, [pageName, CHARTS, chartName], {})
)((state, pageName, chartName) => `${pageName}|${chartName}`)

export const selectPageFactTypes = (state, pageName) =>
    selectDomainValue(state, [pageName, FILTERS, FACT_TYPES])

export const selectPageTimeseriesData = (state, pageName) =>
    selectDomainValue(state, [pageName, 'timeseries'])

export const selectDataForStackedBarChart = createCachedSelector(
    selectPageTimeseriesData,
    selectPageAggregation,
    selectChartMetrics,
    (data, aggregate, chartMetrics) => {
        // Metrics selected for the chart
        const metrics = zipObject(
            CHART_METRIC_KEYS,
            CHART_METRIC_KEYS.map(chartMetricKey =>
                getMetric(chartMetrics[chartMetricKey])
            )
        )

        const metricAggregates = zipObject(
            FACT_TYPE_KEYS,
            FACT_TYPE_KEYS.map(factTypeKey =>
                zipObject(
                    CHART_METRIC_KEYS,
                    CHART_METRIC_KEYS.map(chartMetricKey =>
                        aggregateMetricByDateGroups({
                            data: data[factTypeKey],
                            metric: chartMetrics[chartMetricKey],
                            aggregate,
                            order: 'asc',
                        })
                    )
                )
            )
        )

        // Generate axes data
        const axes = {
            date: {
                tickInterval: moment.duration(1, aggregate).asMilliseconds(),
                tickFormat: value => {
                    const formattedDate = formatDate(value, aggregate)
                    return aggregate === 'week'
                        ? `
                        <span>
                            Wk. End <br />
                            ${formattedDate}
                        </span>
                    `
                        : formattedDate
                },
            },
            left: {
                tickFormat: value => {
                    const { format } = metrics[LEFT_AXIS_FIRST]

                    return formatNumber(value, format)
                },
            },
            right: {
                tickFormat: value => {
                    const { format } = metrics[RIGHT_AXIS]

                    return formatNumber(value, format)
                },
            },
        }

        // Generate series data
        const series = zipObject(
            FACT_TYPE_KEYS,
            FACT_TYPE_KEYS.map(factTypeKey =>
                zipObject(
                    CHART_METRIC_KEYS,
                    CHART_METRIC_KEYS.map(chartMetricKey => {
                        if (
                            isEmpty(
                                metricAggregates[factTypeKey][chartMetricKey]
                            )
                        ) {
                            return []
                        }

                        return metricAggregates[factTypeKey][
                            chartMetricKey
                        ].map(value => [
                            aggregate === 'week'
                                ? moment(value.period_end)
                                      .startOf('day')
                                      .valueOf()
                                : moment(value.period_start).valueOf(),
                            value[chartMetrics[chartMetricKey]],
                        ])
                    })
                )
            )
        )

        return {
            axes,
            series,
        }
    }
)((state, pageName, chartName) => `${pageName}|${chartName}`)

export const selectLoadingForStackedBarChart = (state, pageName) =>
    selectDomainValue(state, [pageName, 'timeseries', 'loading'])
