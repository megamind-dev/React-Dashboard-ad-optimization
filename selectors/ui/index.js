export * from './ui'
export * from './labels'
export * from './globalNotification'
export * from './filters'
export * from './tables'
export * from './amsTreemapCharts'
export * from './amsTimeseriesCharts'
export * from './sovCharts'
export * from './organicSearchPositionCharts'
export * from './page'
