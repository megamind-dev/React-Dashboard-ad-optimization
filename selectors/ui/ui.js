import createCachedSelector from 're-reselect'
import get from 'lodash/get'

export const selectDomainState = state => state.ui

export const selectDomainValue = createCachedSelector(
    selectDomainState,
    (state, path) => path,
    (domainState, path) => get(domainState, path)
)((state, path) => JSON.stringify(path))
