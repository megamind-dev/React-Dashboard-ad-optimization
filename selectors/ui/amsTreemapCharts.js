import { createSelector } from 'reselect'
import get from 'lodash/get'
import flow from 'lodash/fp/flow'
import filter from 'lodash/fp/filter'
import orderBy from 'lodash/fp/orderBy'
import map from 'lodash/fp/map'

import { formatNumber } from 'helpers/formatting'
import { metrics } from 'configuration/metrics'

import { selectDomainValue } from './ui'

export const selectTreemapSelectedMetric = (state, pageName, treemapName) =>
    selectDomainValue(state, [pageName, treemapName, 'sorter', 'field'])

export const selectTreemap = (state, pageName, treemapName) =>
    selectDomainValue(state, [pageName, treemapName])

export const selectTreemapChartData = createSelector(
    selectTreemap,
    ({ data, sorter }) =>
        flow(
            filter(row => get(row, sorter.field, 0) > 0),
            orderBy([sorter.field], ['desc']), // Always order in descending order
            map(row => {
                const metricMetadata = metrics[sorter.field]
                const value = row[sorter.field]

                return {
                    metric: {
                        nativeValue: value,
                        formattedValue: formatNumber(
                            value,
                            metricMetadata.format
                        ),
                    },
                    metricMetadata,
                    resourceData: row,
                }
            })
        )(data)
)
