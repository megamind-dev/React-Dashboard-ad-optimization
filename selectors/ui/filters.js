import createCachedSelector from 're-reselect'
import get from 'lodash/get'
import pickBy from 'lodash/pickBy'

import { FILTERS, FILTER_SETTINGS } from 'constants/reducerKeys'
import { AGGREGATION } from 'constants/filters'

import { selectDomainState } from './ui'

export const selectPageFilters = createCachedSelector(
    selectDomainState,
    (state, pageName) => pageName,
    (uiState, pageName) => get(uiState, [pageName, FILTERS], {})
)((state, pageName) => pageName)

export const selectPageFilterSettings = createCachedSelector(
    selectDomainState,
    (state, pageName) => pageName,
    (uiState, pageName) => get(uiState, [pageName, FILTER_SETTINGS], {})
)((state, pageName) => pageName)

export const selectVisiblePageFilters = createCachedSelector(
    (state, pageName) => selectPageFilters(state, pageName),
    (state, pageName) => selectPageFilterSettings(state, pageName),
    (filters, { anchored = [], displayState }) =>
        pickBy(
            filters,
            (value, key) => anchored.includes(key) || displayState[key]
        )
)((state, pageName) => pageName)

export const selectPageAggregation = createCachedSelector(
    selectDomainState,
    (state, pageName) => pageName,
    (domainState, pageName) =>
        get(domainState, [pageName, FILTERS, AGGREGATION])
)((state, pageName) => pageName)
