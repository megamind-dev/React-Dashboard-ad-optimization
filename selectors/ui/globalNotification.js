import { selectDomainValue } from './ui'

export const selectGlobalNotification = state =>
    selectDomainValue(state, ['app', 'globalNotification'])
