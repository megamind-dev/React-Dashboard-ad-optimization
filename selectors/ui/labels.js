import { createSelector } from 'reselect'
import map from 'lodash/fp/map'
import flow from 'lodash/flow'
import orderBy from 'lodash/fp/orderBy'
import { selectDomainValue } from './ui'

const selectLabels = state =>
    selectDomainValue(state, ['app', 'organizationLabels'])

export const selectLabelsForFilters = createSelector(
    selectLabels,
    labels =>
        flow(
            map(label => ({
                label: label.name,
                value: label.id,
            })),
            orderBy(label => label.name, 'asc')
        )(labels)
)
