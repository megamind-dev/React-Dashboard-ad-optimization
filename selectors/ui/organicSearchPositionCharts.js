import createCachedSelector from 're-reselect'
import map from 'lodash/fp/map'
import flow from 'lodash/flow'
import orderBy from 'lodash/fp/orderBy'
import moment from 'utilities/moment'

import { ORGANIC_SEARCH_POSITION_CHART } from 'constants/reducerKeys'
import { SOV_KEYWORD_ORGANIC_SEARCH_POSITION_KEY } from 'constants/sovKeywords'
import { SOV_AGGREGATION } from 'constants/filters'

import { selectDomainValue } from './ui'
import { selectPageFilters } from './filters'

const selectPageOrganicSearchPositionChartData = (state, pageName) =>
    selectDomainValue(state, [pageName, ORGANIC_SEARCH_POSITION_CHART, 'data'])

export const selectDataForOrganicSearchPositionChart = createCachedSelector(
    selectPageOrganicSearchPositionChartData,
    selectPageFilters,
    (chartData, filters) => {
        const aggregate = filters[SOV_AGGREGATION]
        const points = flow(
            map(result => {
                const yValue = result[SOV_KEYWORD_ORGANIC_SEARCH_POSITION_KEY]
                return {
                    x: moment(result[aggregate]).valueOf(),
                    y: yValue !== -1 ? yValue + 1 : null,
                }
            }),
            orderBy(point => point.x, 'asc')
        )(chartData)
        const series = [
            {
                name: 'Organic Search Position',
                data: points,
            },
        ]

        return { series }
    }
)((state, pageName) => pageName)

export const selectSelectedAsinForOrganicSearchPositionChart = (
    state,
    pageName
) =>
    selectDomainValue(state, [
        pageName,
        ORGANIC_SEARCH_POSITION_CHART,
        'selectedAsin',
    ])

export const selectLoadingForOrganicSearchPositionChart = (state, pageName) =>
    selectDomainValue(state, [
        pageName,
        ORGANIC_SEARCH_POSITION_CHART,
        'loading',
    ])
