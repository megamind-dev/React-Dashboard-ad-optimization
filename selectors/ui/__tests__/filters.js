import { HOME_PAGE } from 'constants/pages'
import reducer from 'reducers'
import { updatePageFilter, updatePageFilterSettings } from 'actions/ui'

import {
    selectPageFilters,
    selectPageFilterSettings,
    selectVisiblePageFilters,
} from '../filters'

describe('[Selectors] ui/filters', () => {
    const pageName = HOME_PAGE
    let state

    beforeEach(() => {
        const initActions = [
            updatePageFilter({
                pageName,
                data: {
                    key: 'countries',
                    value: ['US', 'CA'],
                },
            }),
            updatePageFilter({
                pageName,
                data: {
                    key: 'regions',
                    value: ['EU'],
                },
            }),
            updatePageFilterSettings({
                pageName,
                data: {
                    displayState: {
                        countries: true,
                        regions: true,
                    },
                },
            }),
        ]

        state = initActions.reduce(
            (tempState, action) => reducer(tempState, action),
            undefined
        )
    })

    describe('selectPageFilters', () => {
        it('selects page filters by page name', () => {
            const filters = selectPageFilters(state, pageName)

            expect(filters).toMatchObject({
                countries: ['US', 'CA'],
                regions: ['EU'],
            })
        })

        it('memoizes page filters by page name', () => {
            const selector = selectPageFilters.getMatchingSelector(
                state,
                pageName
            )

            // Reset computations
            selector.resetRecomputations()

            // Repeat computations without state change
            selectPageFilters(state, pageName)
            selectPageFilters(state, pageName)
            expect(selector.recomputations()).toEqual(1)

            // Update state
            state = reducer(
                state,
                updatePageFilter({
                    pageName: HOME_PAGE,
                    data: {
                        key: 'regions',
                        value: ['NA'],
                    },
                })
            )

            // Trigger computation after state change
            selectPageFilters(state, pageName)
            expect(selector.recomputations()).toEqual(2)
        })
    })

    describe('selectPageFilterSettings', () => {
        it('selects page filter settings by page name', () => {
            const filterSettings = selectPageFilterSettings(state, pageName)

            expect(filterSettings).toMatchSnapshot()
        })

        it('memoizes page filter settings by page name', () => {
            const selector = selectPageFilterSettings.getMatchingSelector(
                state,
                pageName
            )

            // Reset computations
            selector.resetRecomputations()

            // Repeat computations without state change
            selectPageFilterSettings(state, pageName)
            selectPageFilterSettings(state, pageName)
            expect(selector.recomputations()).toEqual(1)

            // Update state
            state = reducer(
                state,
                updatePageFilterSettings({
                    pageName: HOME_PAGE,
                    data: {
                        displayState: {
                            regions: false,
                        },
                    },
                })
            )

            // Trigger computation after state change
            selectPageFilterSettings(state, pageName)
            expect(selector.recomputations()).toEqual(2)
        })
    })

    describe('selectVisiblePageFilters', () => {
        it('selects visible page filters by page name', () => {
            const visibleFilters = selectVisiblePageFilters(state, pageName)

            expect(visibleFilters).toMatchObject({
                countries: ['US', 'CA'],
                regions: ['EU'],
            })
        })

        it('memoizes visible page filters by page name', () => {
            const selector = selectVisiblePageFilters.getMatchingSelector(
                state,
                pageName
            )

            // Reset computations
            selector.resetRecomputations()

            // Repeat computations without state change
            selectVisiblePageFilters(state, pageName)
            selectVisiblePageFilters(state, pageName)
            expect(selector.recomputations()).toEqual(1)

            // Update state
            state = reducer(
                state,
                updatePageFilter({
                    pageName: HOME_PAGE,
                    data: {
                        key: 'regions',
                        value: ['NA'],
                    },
                })
            )

            // Trigger computation after state change
            selectVisiblePageFilters(state, pageName)
            expect(selector.recomputations()).toEqual(2)
        })

        describe('when fiters are hidden', () => {
            beforeEach(() => {
                state = reducer(
                    state,
                    updatePageFilterSettings({
                        pageName,
                        data: {
                            displayState: {
                                regions: false,
                            },
                        },
                    })
                )
            })

            it('omits hidden filters', () => {
                const visibleFilters = selectVisiblePageFilters(state, pageName)

                expect(visibleFilters).not.toHaveProperty('regions')
            })
        })
    })
})
