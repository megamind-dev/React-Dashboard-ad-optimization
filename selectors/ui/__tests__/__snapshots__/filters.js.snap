// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[Selectors] ui/filters selectPageFilterSettings selects page filter settings by page name 1`] = `
Object {
  "anchored": Array [
    "aggregation",
    "dates",
    "factTypes",
  ],
  "displayState": Object {
    "brands": true,
    "countries": true,
    "regions": true,
  },
  "order": Array [
    "regions",
    "countries",
    "brands",
  ],
}
`;
